File: src-tauri/src/ipc.rs
/**
 * IPC message types — Rust mirror of contracts/ipc.ts.
 *
 * TS source of truth: contracts/ipc.ts
 * Python mirror:      sidecar/ipc_types.py
 *
 * Framing: [4-byte u32 LE length][UTF-8 JSON payload]
 */
use serde::{Deserialize, Serialize};

// ── Lifecycle ─────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum LifecycleStatus {
    Starting,
    Ready,
    Degraded,
    Error,
    Stopped,
}

impl Default for LifecycleStatus {
    fn default() -> Self {
        Self::Stopped
    }
}

impl std::fmt::Display for LifecycleStatus {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = match self {
            Self::Starting => "starting",
            Self::Ready => "ready",
            Self::Degraded => "degraded",
            Self::Error => "error",
            Self::Stopped => "stopped",
        };
        write!(f, "{s}")
    }
}

// ── Allowlisted actions ───────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "action", rename_all = "snake_case")]
pub enum Action {
    OpenApp {
        app_name: String,
    },
    OpenUrl {
        url: String,
    },
    YoutubeSearch {
        query: String,
    },
    YoutubePlay {
        query: String,
    },
    SpotifyOpen,
    SpotifySearch {
        query: String,
    },
    SpotifyPlay {
        query: String,
    },
    MediaPlay,
    MediaNext,
    MediaPrev,
    ShowSystemSummary,
    Respond {
        input: String,
    },
    // Task 3 — hardware inventory + model advisor
    GetInventory,
    GetAdvisor,
    PullModel {
        model_name: String,
    },
    /// Resolves a chat model that is actually installed locally.
    GetDefaultModel,
    // Task 4 — host-internal chat (Rust forwards to sidecar; never executes directly)
    Chat {
        /// Omitted means "use whatever chat model is installed".
        #[serde(default)]
        model: Option<String>,
        message: String,
    },
    // Host-internal cancel; excluded from AssistantAction / broker enum
    CancelChat {
        request_id: String,
    },
    // Voice
    StartListen {
        #[serde(default)]
        model: Option<String>,
    },
    StopListen,
    CancelVoice {
        #[serde(default)]
        request_id: Option<String>,
    },
    SetMute {
        muted: bool,
    },
    // Wake word
    WakeStatus,
    WakeStart,
    WakeStop,
    WakeConfigure {
        #[serde(default)]
        sensitivity: Option<f64>,
        #[serde(default)]
        cooldown_secs: Option<f64>,
        #[serde(default)]
        phrase: Option<String>,
    },
    // Memory
    MemoryStatus,
    MemoryList,
    MemoryAdd {
        text: String,
        #[serde(default)]
        source: Option<String>,
    },
    MemoryDelete {
        id: i64,
    },
    MemoryClear,
    MemoryClearSession,
    MemorySetEnabled {
        enabled: bool,
    },
    MemoryExport,
    // Typed placeholders — not yet implemented
    VoiceTranscribe {
        _placeholder: bool,
    },
    WakeWordDetected {
        _placeholder: bool,
    },
}

// ── Task 4: Assistant action (Rust broker input) ───────────────────────────────
// Intentionally excludes chat, respond, inventory, pull, and placeholders.
// React MUST NOT pass those variants to execute_assistant_action.

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "action", rename_all = "snake_case")]
pub enum AssistantAction {
    OpenApp { app_name: String },
    OpenUrl { url: String },
    YoutubeSearch { query: String },
    YoutubePlay { query: String },
    SpotifyOpen,
    SpotifySearch { query: String },
    SpotifyPlay { query: String },
    MediaPlay,
    MediaNext,
    MediaPrev,
    ShowSystemSummary,
}

// ── Task 4: Audit record ──────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AuditOutcome {
    Ok,
    Error,
}

/// Append-only audit event emitted to UI after every broker action.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuditEvent {
    /// Unique ID: "audit-<unix_ms>-<counter>"
    pub id: String,
    /// Action kind: open_app | open_url | youtube_* | spotify_* | show_system_summary
    pub action_kind: String,
    /// Sanitized display label (no secret/full-URL leakage).
    pub target_label: String,
    /// Unix epoch milliseconds as decimal string.
    pub timestamp: String,
    pub outcome: AuditOutcome,
    pub error_msg: Option<String>,
}

// ── Host → Sidecar ────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum HostMessage {
    Action { id: String, payload: Action },
    Shutdown,
}

// ── Sidecar → Host ────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum SidecarMessage {
    Ready {
        version: String,
    },
    Response {
        id: String,
        result: String,
    },
    Error {
        id: String,
        error: String,
    },
    Stream {
        id: String,
        chunk: String,
        finished: bool,
    },
    Crash {
        code: i32,
        message: String,
    },
}

// ── Frontend event payload ────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize)]
#[serde(tag = "event", rename_all = "kebab-case")]
pub enum AppEventPayload {
    LifecycleChanged {
        lifecycle: LifecycleStatus,
        reason: Option<String>,
    },
    SidecarMessage {
        message: SidecarMessage,
    },
    CrashReport {
        pid: u32,
        exit_code: i32,
        crash_count: u32,
    },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn lifecycle_status_roundtrip() {
        let statuses = [
            LifecycleStatus::Starting,
            LifecycleStatus::Ready,
            LifecycleStatus::Degraded,
            LifecycleStatus::Error,
            LifecycleStatus::Stopped,
        ];
        for s in statuses {
            let json = serde_json::to_string(&s).unwrap();
            let back: LifecycleStatus = serde_json::from_str(&json).unwrap();
            assert_eq!(s, back, "round-trip failed for {s:?}");
        }
    }

    #[test]
    fn host_message_shutdown_serializes_correctly() {
        let msg = HostMessage::Shutdown;
        let json = serde_json::to_string(&msg).unwrap();
        assert!(json.contains("\"type\":\"shutdown\""), "got: {json}");
    }

    #[test]
    fn sidecar_ready_message_roundtrip() {
        let msg = SidecarMessage::Ready {
            version: "0.1.0".to_string(),
        };
        let json = serde_json::to_string(&msg).unwrap();
        let back: SidecarMessage = serde_json::from_str(&json).unwrap();
        if let SidecarMessage::Ready { version } = back {
            assert_eq!(version, "0.1.0");
        } else {
            panic!("wrong variant after round-trip");
        }
    }

    #[test]
    fn action_open_app_serializes_tag() {
        let a = Action::OpenApp {
            app_name: "notepad".to_string(),
        };
        let json = serde_json::to_string(&a).unwrap();
        assert!(json.contains("\"action\":\"open_app\""), "got: {json}");
        assert!(json.contains("notepad"), "got: {json}");
    }

    #[test]
    fn action_get_inventory_serializes_tag() {
        let json = serde_json::to_string(&Action::GetInventory).unwrap();
        assert!(json.contains("\"action\":\"get_inventory\""), "got: {json}");
    }

    #[test]
    fn action_get_advisor_serializes_tag() {
        let json = serde_json::to_string(&Action::GetAdvisor).unwrap();
        assert!(json.contains("\"action\":\"get_advisor\""), "got: {json}");
    }

    #[test]
    fn action_pull_model_serializes_tag() {
        let a = Action::PullModel {
            model_name: "llama3.2:1b-instruct-q4_K_M".to_string(),
        };
        let json = serde_json::to_string(&a).unwrap();
        assert!(json.contains("\"action\":\"pull_model\""), "got: {json}");
        assert!(json.contains("llama3.2"), "got: {json}");
    }

    #[test]
    fn action_chat_serializes_tag() {
        let a = Action::Chat {
            model: Some("llama3.2:1b".to_string()),
            message: "hello".to_string(),
        };
        let json = serde_json::to_string(&a).unwrap();
        assert!(json.contains("\"action\":\"chat\""), "got: {json}");
        assert!(json.contains("\"model\""), "got: {json}");
        assert!(json.contains("\"message\""), "got: {json}");
    }

    #[test]
    fn assistant_action_open_url_roundtrip() {
        let a = AssistantAction::OpenUrl {
            url: "https://example.com".to_string(),
        };
        let json = serde_json::to_string(&a).unwrap();
        assert!(json.contains("\"action\":\"open_url\""), "got: {json}");
        let back: AssistantAction = serde_json::from_str(&json).unwrap();
        if let AssistantAction::OpenUrl { url } = back {
            assert_eq!(url, "https://example.com");
        } else {
            panic!("wrong variant after roundtrip");
        }
    }

    #[test]
    fn assistant_action_open_app_roundtrip() {
        let a = AssistantAction::OpenApp {
            app_name: "Notepad".to_string(),
        };
        let json = serde_json::to_string(&a).unwrap();
        assert!(json.contains("\"action\":\"open_app\""), "got: {json}");
        let back: AssistantAction = serde_json::from_str(&json).unwrap();
        if let AssistantAction::OpenApp { app_name } = back {
            assert_eq!(app_name, "Notepad");
        } else {
            panic!("wrong variant after roundtrip");
        }
    }

    #[test]
    fn assistant_action_rejects_chat_variant() {
        // Rust broker enum must not deserialize "chat"
        let json = r#"{"action":"chat","model":"x","message":"y"}"#;
        let result: Result<AssistantAction, _> = serde_json::from_str(json);
        assert!(result.is_err(), "AssistantAction must not accept 'chat'");
    }

    #[test]
    fn assistant_action_rejects_get_inventory() {
        let json = r#"{"action":"get_inventory"}"#;
        let result: Result<AssistantAction, _> = serde_json::from_str(json);
        assert!(
            result.is_err(),
            "AssistantAction must not accept 'get_inventory'"
        );
    }

    #[test]
    fn action_cancel_chat_serializes_tag() {
        let a = Action::CancelChat {
            request_id: "req-123".to_string(),
        };
        let json = serde_json::to_string(&a).unwrap();
        assert!(json.contains("\"action\":\"cancel_chat\""), "got: {json}");
        assert!(json.contains("\"request_id\":\"req-123\""), "got: {json}");
    }

    #[test]
    fn assistant_action_rejects_cancel_chat() {
        let json = r#"{"action":"cancel_chat","request_id":"x"}"#;
        let result: Result<AssistantAction, _> = serde_json::from_str(json);
        assert!(
            result.is_err(),
            "AssistantAction must not accept 'cancel_chat'"
        );
    }

    #[test]
    fn audit_event_serializes_correctly() {
        let ev = AuditEvent {
            id: "audit-1-0".to_string(),
            action_kind: "open_url".to_string(),
            target_label: "example.com".to_string(),
            timestamp: "1700000000000".to_string(),
            outcome: AuditOutcome::Ok,
            error_msg: None,
        };
        let json = serde_json::to_string(&ev).unwrap();
        assert!(json.contains("\"outcome\":\"ok\""), "got: {json}");
        assert!(json.contains("\"action_kind\":\"open_url\""), "got: {json}");
    }
}

File: contracts/ipc.ts
/**
 * Canonical IPC contract — single source of truth for message shapes.
 *
 * Rust mirror  : src-tauri/src/ipc.rs
 * Python mirror: sidecar/ipc_types.py
 *
 * All three representations MUST stay in sync.
 * Protocol: [4-byte u32 LE length][UTF-8 JSON payload]
 */

// ── Lifecycle ─────────────────────────────────────────────────────────────────

/** App / sidecar lifecycle states. */
export type AppLifecycle =
  | "starting"   // Sidecar process spawned, awaiting ready handshake
  | "ready"      // Sidecar sent { type: "ready" }
  | "degraded"   // Sidecar running but a crash was recovered; functionality partial
  | "error"      // Unrecoverable state; user action required
  | "stopped";   // Sidecar not running (initial or after clean shutdown)

// ── Allowlisted actions (MVP) ─────────────────────────────────────────────────

export type Action =
  | { action: "open_app";           app_name: string }
  | { action: "open_url";           url: string }
  | { action: "youtube_search";     query: string }
  | { action: "youtube_play";       query: string }
  | { action: "spotify_open" }
  | { action: "spotify_search";     query: string }
  | { action: "spotify_play";       query: string }
  | { action: "media_play" }
  | { action: "media_next" }
  | { action: "media_prev" }
  | { action: "show_system_summary" }
  | { action: "respond";            input: string }
  // Task 3 — hardware inventory + model advisor
  | { action: "get_inventory" }
  | { action: "get_advisor" }
  | { action: "pull_model";         model_name: string }
  /** Resolves a chat model that is actually installed locally. */
  | { action: "get_default_model" }
  // Task 4 — typed assistant chat (host-internal; Rust only forwards, never executes)
  | { action: "chat"; model?: string; message: string }
  // Host-internal cancel; excluded from AssistantAction / broker enum
  | { action: "cancel_chat"; request_id: string }
  // Voice
  | { action: "start_listen"; model?: string }
  | { action: "stop_listen" }
  | { action: "cancel_voice"; request_id?: string }
  | { action: "set_mute"; muted: boolean }
  // Wake word
  | { action: "wake_status" }
  | { action: "wake_start" }
  | { action: "wake_stop" }
  | { action: "wake_configure"; sensitivity?: number; cooldown_secs?: number; phrase?: string; profile?: string }
  // Memory
  | { action: "memory_status" }
  | { action: "memory_list" }
  | { action: "memory_add"; text: string; source?: string }
  | { action: "memory_delete"; id: number }
  | { action: "memory_clear" }
  | { action: "memory_clear_session" }
  | { action: "memory_delete_session"; id: number }
  | { action: "memory_set_enabled"; enabled: boolean }
  | { action: "memory_export" }
  // Screen context (opt-in focused-window text)
  | { action: "screen_status" }
  | { action: "screen_set_enabled"; enabled: boolean }
  | { action: "get_focused_window_text" }
  // Browser tools
  | { action: "browser_scroll"; direction: string; steps?: number }
  | { action: "browser_type"; text: string }
  | { action: "browser_click_role"; role?: string; name: string }
  | { action: "browser_focus_search" }
  | { action: "browser_confirm"; pending_id: string }
  | { action: "browser_cancel"; pending_id: string }
  // Typed placeholders — not yet implemented
  | { action: "voice_transcribe";   _placeholder: true }
  | { action: "wake_word_detected"; _placeholder: true };

// ── Task 4: Assistant action types ───────────────────────────────────────────
// Subset of Action that the Rust broker can execute.
// Intentionally excludes: chat, respond, get_inventory, get_advisor, pull_model, placeholders.

export type AssistantAction =
  | { action: "open_app";           app_name: string }
  | { action: "open_url";           url: string }
  | { action: "youtube_search";     query: string }
  | { action: "youtube_play";       query: string }
  | { action: "spotify_open" }
  | { action: "spotify_search";     query: string }
  | { action: "spotify_play";       query: string }
  | { action: "media_play" }
  | { action: "media_next" }
  | { action: "media_prev" }
  | { action: "show_system_summary" };

/** Typed result returned by Python sidecar after a chat turn. */
export type AssistantResult =
  | { kind: "respond"; text: string }
  | { kind: "action";  action: AssistantAction };

/** Append-only audit record emitted by the Rust broker after executing an action. */
export interface AuditEvent {
  /** Monotonic unique ID (e.g. "audit-<ms>-<n>"). */
  id: string;
  /** Allowlisted broker action kind (open_app, youtube_*, spotify_*, …). */
  action_kind: string;
  /** Sanitized display label (no secret/full-URL leakage). */
  target_label: string;
  /** Unix epoch milliseconds as decimal string. */
  timestamp: string;
  outcome: "ok" | "error";
  /** Human-readable error message; null on success. */
  error_msg: string | null;
}

// ── Task 3: Inventory result types ───────────────────────────────────────────
// Returned as JSON-encoded string in SidecarMessage.Response.result

export interface GpuInfo { name: string; vram_gb: number }

export interface HardwareInfo {
  os: string;
  cpu: string;
  ram_gb: number;
  gpu: GpuInfo | null;
  /** Empty string when GPU was detected; NVIDIA-only disclosure otherwise.
   *  Never treat null gpu as "no GPU present" — AMD/Intel/other GPUs are
   *  simply not detected by nvidia-smi. */
  gpu_note: string;
  mic_available: boolean;
}

export interface OllamaModel {
  name: string;
  size_gb: number;
  quantization: string | null;
}

export interface OllamaStatus {
  reachable: boolean;
  version: string | null;
  models: OllamaModel[];
  running: string[];
}

export interface InstalledApp {
  name: string;
  source: "start_menu" | "registry";
}

export interface InventoryResult {
  hardware: HardwareInfo;
  ollama: OllamaStatus;
  apps: InstalledApp[];
}

// ── Task 3: Advisor result types ─────────────────────────────────────────────

export type AdvisorTier       = "fast" | "balanced" | "quality";
export type AdvisorConstraint = "cpu_only" | "vram_limited" | "vram_ok";

export interface AdvisorRecommendation {
  tier:           AdvisorTier;
  candidate_name: string;
  display_name:   string;
  size_gb:        number;
  context_k:      number;
  quantization:   string;
  reason:         string;
  available:      boolean;
}

export interface AdvisorResult {
  catalog_version:  string;
  recommendations:  AdvisorRecommendation[];
  constraint:       AdvisorConstraint;
  warning:          string | null;
}

/** Combined response for the get_advisor action. */
export interface GetAdvisorResponse {
  hardware: HardwareInfo;
  ollama:   OllamaStatus;
  advisor:  AdvisorResult;
}

// ── Host → Sidecar (stdin) ────────────────────────────────────────────────────

export type HostMessage =
  | { type: "action";   id: string; payload: Action }
  | { type: "shutdown" };

// ── Sidecar → Host (stdout) ───────────────────────────────────────────────────

export type SidecarMessage =
  | { type: "ready";    version: string }
  | { type: "response"; id: string; result: string }
  | { type: "error";    id: string; error: string }
  | { type: "stream";   id: string; chunk: string; finished: boolean }
  | { type: "crash";    code: number; message: string };

// ── Tauri → Frontend events ───────────────────────────────────────────────────

export type AppEvent =
  | { event: "lifecycle-changed"; lifecycle: AppLifecycle; reason: string | null }
  | { event: "sidecar-message";   message: SidecarMessage }
  | { event: "crash-report";      pid: number; exit_code: number; crash_count: number };

/** Emitted on the "audit-event" channel (separate from "app-event"). */
export type AuditAppEvent = AuditEvent;
